bugbug/tools/release_notes.py

```diff
@@ -0,0 +1,217 @@
+- Bug fixes that don’t impact most users.
+- Obscure web compatibility changes that apply only to edge-case websites.
+- Duplicate entries or similar changes that were already listed.
+
+Here is the list to filter:
+{combined_list}
+
+Instructions:
+- KEEP THE SAME FORMAT (do not change the structure of entries that remain).
+- REMOVE UNWORTHY ENTRIES ENTIRELY (do not rewrite them—just delete).
+- DO NOT ADD ANY TEXT BEFORE OR AFTER THE LIST.
+- The output must be only the cleaned-up list, formatted exactly the same way.
+""",
+        )
+
+        self.cleanup_chain = LLMChain(
+            llm=self.llm,
+            prompt=self.cleanup_prompt,
+        )
+
+    def get_previous_version(self, current_version: str) -> str:
+        match = re.match(r"(FIREFOX_BETA_)(\d+)(_BASE)", current_version)
+        if not match:
+            raise ValueError("Invalid version format")
+        prefix, version_number, suffix = match.groups()
+        previous_version_number = int(version_number) - 1
+        return f"{prefix}{previous_version_number}{suffix}"
+
+    def get_token_count(self, text: str) -> int:
+        if hasattr(self.llm, "model_name"):
+            model_name = self.llm.model_name
+        else:
+            raise ValueError("LLM model name not found.")
+
+        encoding = tiktoken.encoding_for_model(model_name)
+        return len(encoding.encode(text))
+
+    def split_into_chunks(self, commit_log: str) -> list[str]:
+        commit_blocks = commit_log.split("\n")
+        chunks = []
+        current_chunk: list[str] = []
+        current_token_count = 0
+
+        for block in commit_blocks:
+            block_token_count = self.get_token_count(block)
+
+            if current_token_count + block_token_count > self.chunk_size:
+                chunks.append("\n\n".join(current_chunk))
+                current_chunk = []
+                current_token_count = 0
+
+            current_chunk.append(block)
+            current_token_count += block_token_count
+
+        if current_chunk:
+            chunks.append("\n\n".join(current_chunk))
+
+        return chunks
+
+    def shortlist_with_gpt(self, input_text: str) -> str:
+        return self.summarization_chain.run({"input_text": input_text}).strip()
+
+    def generate_commit_shortlist(self, commit_log_list: list[str]) -> list[str]:
+        commit_log_list_combined = "\n".join(commit_log_list)
+        chunks = self.split_into_chunks(commit_log_list_combined)
+        return [self.shortlist_with_gpt(chunk) for chunk in chunks]
+
+    def clean_commits(
+        self, commit_log_list: list[str], keywords: list[str]
+    ) -> list[str]:
+        cleaned_commits = []
+
+        for block in commit_log_list:
+            if (
+                not any(
+                    re.search(rf"\b{keyword}\b", block, re.IGNORECASE)
+                    for keyword in keywords
+                )
+                and re.search(r"Bug \d+", block, re.IGNORECASE)
```

Comment:
I do think this is needed since we are already changing that later in `if not bug_id_match`.